nexus/src/db/datastore/region.rs

```diff
@@ -127,204 +123,41 @@ impl DataStore {
-                        .filter(dataset_dsl::pool_id.eq(zpool_id))
-                        .filter(dataset_dsl::size_used.is_not_null())
-                        .filter(dataset_dsl::time_deleted.is_null())
-                        .select(diesel::dsl::sum(dataset_dsl::size_used))
-                        .nullable()
-                        .get_result_async(&conn)
-                        .await?;
-
-                    let zpool_total_occupied_size: u64 = if let Some(
-                        zpool_total_occupied_size,
-                    ) =
-                        zpool_total_occupied_size
-                    {
-                        let zpool_total_occupied_size: db::model::ByteCount =
-                            zpool_total_occupied_size.try_into().map_err(
-                                |e: anyhow::Error| {
-                                    TxnError::CustomError(
-                                        RegionAllocateError::NumericError(
-                                            e.to_string(),
-                                        ),
-                                    )
-                                },
-                            )?;
+        let mut regions: Vec<Region> =
+            crate::db::queries::region_allocation::RegionAllocate::new(
+                volume_id,
+                block_size.into(),
+                blocks_per_extent,
+                params.extent_count(),
+            )
+            .get_results_async(self.pool())
+            .await
+            .map_err(|e| crate::db::queries::region_allocation::from_pool(e))?;
 
-                        zpool_total_occupied_size.to_bytes()
-                    } else {
-                        0
-                    };
+        let dataset_ids: Vec<Uuid> =
+            regions.iter().map(|r| r.dataset_id()).collect();
 
-                    let zpool = zpool_dsl::zpool
-                        .filter(zpool_dsl::id.eq(zpool_id))
-                        .select(Zpool::as_returning())
-                        .get_result_async(&conn)
-                        .await?;
+        let mut datasets: Vec<Dataset> = dataset_dsl::dataset
+            .filter(dataset_dsl::id.eq_any(dataset_ids))
+            .select(Dataset::as_select())
+            .get_results_async(self.pool())
+            .await
+            .map_err(|e| {
+                public_error_from_diesel_pool(e, ErrorHandler::Server)
+            })?;
 
-                    // Does this go over the zpool's total size?
-                    if zpool.total_size.to_bytes() < zpool_total_occupied_size {
-                        return Err(TxnError::CustomError(
-                            RegionAllocateError::NotEnoughAvailableSpace,
-                        ));
-                    }
-                }
+        datasets.sort_by(|a, b| a.id().partial_cmp(&b.id()).unwrap());
+        regions.sort_by(|a, b| {
+            a.dataset_id().partial_cmp(&b.dataset_id()).unwrap()
+        });
```

Comment:
Done in https://github.com/oxidecomputer/omicron/pull/1733/commits/653a20b0722b9827a755f78924b100ce7edf5ce4 - it was unfortunately more complicated than your proposal.

The problem is that your SQL statement, as written, returns the *old* value of datasets, and those are potentially updated by the CTE too. In other words, this fails `db::datastore::test::test_region_allocation_is_idempotent`, because the datasets would appear different on the second query.

I've done "things" some people might consider "kinda gnarly", but pulled it off via a double-join thrown together with a UNION.